lib/server/services/qualityControlFlag/QcFlagService.js

```diff
@@ -301,7 +406,78 @@ class QcFlagService {
+                ],
+                sort: [['createdAt', 'ASC']],
+            });
+
+            const flagsCreatedBeforeRemovedFlag = await QcFlagRepository.findAll({
+                where: {
+                    dplDetectorId: removedFlag.dplDetectorId,
+                    runNumber: removedFlag.runNumber,
+                    createdAt: { [Op.lte]: qcFlag.createdAt.getTime() },
+                },
+                include: [
+                    qcFlag.dataPasses?.length > 0
+                        ? { association: 'dataPasses', where: { id: qcFlag.dataPasses[0].id } }
+                        : { association: 'simulationPasses', where: { id: qcFlag.simulationPasses[0].id } },
+                ],
+                sort: [['createdAt', 'ASC']],
+            });
+            {
+                await QcFlagEffectivePeriodRepository.removeAll({
+                    where: { flagId: { [Op.in]: flagsCreatedBeforeRemovedFlag.map(({ id }) => id) } },
+                });
+                await QcFlagEffectivePeriodRepository.insertAll(flagsCreatedBeforeRemovedFlag
+                    .map(({ id, from, to }) => ({ flagId: id, from, to })));
+            }
+
+            for (const index in flagsCreatedBeforeRemovedFlag) {
+                const olderUpdatingFlags = flagsCreatedBeforeRemovedFlag.slice(Number(index) + 1);
+                for (const updatingFlag of [...olderUpdatingFlags, ...flagsCreatedAfterRemovedFlag]) {
+                    const effectivePeriods = await QcFlagEffectivePeriodRepository.findAll({
+                        where: {
+                            from: { [Op.lt]: updatingFlag.to },
+                            to: { [Op.gt]: updatingFlag.from },
+                        },
+                        include: [
+                            {
+                                association: 'flag',
+                                include: [
+                                    qcFlag.dataPasses?.length > 0
+                                        ? { association: 'dataPasses', where: { id: qcFlag.dataPasses[0].id } }
+                                        : { association: 'simulationPasses', where: { id: qcFlag.simulationPasses[0].id } },
+                                ],
+                                where: {
+                                    [Op.and]: [
+                                        {
+                                            dplDetectorId: updatingFlag.dplDetectorId,
+                                            runNumber: updatingFlag.runNumber,
+                                        },
+                                        { createdAt: { [Op.lt]: updatingFlag.createdAt } },
+                                    ],
+                                },
+                            },
+                        ],
+                    });
+                    await this._updateEffectivePeriods(updatingFlag, effectivePeriods);
+                }
+            }
```

Comment:
I think this part can be simplified: for me, the only effective periods that may be increased are the ones that:
- ends where the deleted effective period starts
- starts where the deleted effective period ends
So basically, I would propose the given algorithm:
- fetch effective periods (for the given QC flag of course) that either start when deleted one stops or ends when the deleted one starts, sorted by createdAt DESC
- store the potential start, stop (start and stop of the effective period being deleted, noted potentialStart and potentialEnd later on)
- loop over the 2 candidate (they might be 0, 1 or 2) and do:
- update effective period start to min(currentStart, Max(potentialStart, currentQcFlagStart)), then potentialStart = Min(potentialStart, the updated value)
- update effective period end to max(currentEnd, Min(potentialEnd, currentQcFlagEnd)), then potentialEnd = Max(potentialEnd, the updated value)

Reply:
As mentioned by Lukasz, this do not handle some cases, closing the conversation